### PR TITLE
feat(testloop): jsonrpc server setup

### PR DESCRIPTION
**Pull request by jancionear**
Add `TestLoopBuilder::with_jsonrpc()`, which starts a real jsonrpc server for each node in the TestLoop.
This will allow us to test some sharded rpc behaviors using TestLoop.

There is a bit of a problem because jsonrpc runs as a real-time async axum app, and TestLoop operates on fake-time actors. We can't connect the real `ViewClientActor` to the jsonrpc server, because it will not respond to a request until testloop advances. To deal with this let's spawn a separate real-time `ViewClientActor` which can be used by the jsonrpc server to query things from the database.

Real-time methods like `tx_status` won't work properly, because the chain doesn't advance in real time, and timeouts might not work as they should.
Submitting transactions is not supported. We could do it by spawning a real-time `RpcHandlerActor`, but that can be added later.

This doesn't handle rosetta rpc, for rosetta we will have to spawn a separate axum service.

Jsonrpc and testloop are a bit incompatible, and connecting them together is not very pretty, but I still think it would be useful for testing sharded rpcs. I'm open to ideas for a better solution.

### CODE REVIEW

File: test-loop-tests/src/utils/rpc.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            Controls::default(),
+            client.validator_signer.clone(),
+        );
+
+        // Build the axum Router.
+        let (_, block_notification_rx) = tokio::sync::watch::channel(None);
+        let rpc_config = RpcConfig {
+            addr: tcp::ListenerAddr::reserve_for_test(),
+            prometheus_addr: None,
+            cors_allowed_origins: vec!["*".to_string()],
```
[the remark this answers is not in the document]
Comment (jancionear) [reply test-loop-tests/src/utils/rpc.rs:64]: Good catch! fixed.

**Comment by Trisfald** [comment]
What do you think about using `axum::Router` directly as `tower::Service` in test loop, instead of spawning an external ViewClientActor? Without any TCP listener.

Setup:
- Build `Router` with senders pointing at the existing TestLoop actors
- Spawn `router.oneshot(http_request)` via `FutureSpawner`
Flow:
- TestLoop polls the future → future sends message to ViewClientActor → TestLoop processes it → response wakes the future → TestLoop polls again → future completes

That bring us to single-threaded deterministic execution, no extra actors, no TCP ports and fake-time control for read-only queries.

`tx_status_fetch` won't work out of the box because of real-time calls. But we could fix that later with injecting JsonRpcHandler with fake-time compatible clocks.

On the other hand, we won't test the HTTP layer in TestLoop (not a big loss), and the RPC operations becomes tightly coupled with the chain advancement (good or bad? idk).

File: test-loop-tests/src/setup/state.rs [hunk 2]
```diff
@@ -101,6 +102,8 @@ pub struct NodeExecutionData {
     /// Set by delay_endorsements_propagation to account for certification delay in timeouts.
     /// It is Arc<_> so updates are visible through clones.
     pub(super) expected_execution_delay: Arc<AtomicU64>,
+    /// JSON-RPC test server for this node. Wrapped in Arc so that NodeExecutionData remains Clone.
+    pub(crate) rpc_setup: Option<Arc<TestLoopRpcServer>>,
```
Comment (Trisfald) [inline test-loop-tests/src/setup/state.rs:106]: AI claims an old issue might become active with this new field member.

in 

```
    pub fn shutdown_and_drain_remaining_events(mut self, timeout: Duration) {
        // State sync dumper is not an Actor, handle stopping separately.
        for node_data in self.node_datas {
            self.test_loop.data.get_mut(&node_data.state_sync_dumper_handle).stop();
        }

        self.test_loop.shutdown_and_drain_remaining_events(timeout);
    }
```

`node_data` are dropped during the loop, and that now drops `TestLoopRpcServer`

It should be safer to use `&self.node_datas` and let destructors run at the end of the method

**Comment by jancionear** [comment]
> What do you think about using axum::Router directly as tower::Service in test loop, instead of spawning an external ViewClientActor? Without any TCP listener.

The tight coupling might be problematic. In the current solution I like that we can put the chain in some specific state (one node a few blocks behind, one syncing, one broken) and run jsonrpc queries on this frozen state to test corner cases. With tight coupling this becomes harder.

There's also an issue of sharded sub-queries - a jsonrpc service might want to make a query to another jsonrpc service on another node. This works out of the box  with a real HTTP server, using raw futures sounds more difficult.

File: test-loop-tests/src/setup/state.rs (hunk 2 above)
Comment (jancionear) [reply test-loop-tests/src/setup/state.rs:106]: The only way I can see it breaking things is if something inside of `shutdown_and_drain_remaining_events` tries to perform jsonrpc queries, which doesn't work because node_datas were dropped.

Tbh this sounds like a stretch, but I can change it to `&self.node_datas` :+1:

**Comment by Trisfald** [comment]
> > What do you think about using axum::Router directly as tower::Service in test loop, instead of spawning an external ViewClientActor? Without any TCP listener.
> 
> The tight coupling might be problematic. In the current solution I like that we can put the chain in some specific state (one node a few blocks behind, one syncing, one broken) and run jsonrpc queries on this frozen state to test corner cases. With tight coupling this becomes harder.
> 
> There's also an issue of sharded sub-queries - a jsonrpc service might want to make a query to another jsonrpc service on another node. This works out of the box with a real HTTP server, using raw futures sounds more difficult.

That's fair concern. 
I was thinking the tight coupling could make implementing `tx_status` easier, but running queries on frozen state is also a good thing to have. Overall I agree we can go ahead with the PR approach

**Comment by shreyan-gupta** [comment]
I'll take a look at this PR today

**Comment by jancionear** [comment]
Closing in favor of https://github.com/near/nearcore/pull/15156
